graphql_compiler/macros/macro_edge/validation.py

```diff
@@ -104,33 +104,29 @@ def _validate_macro_edge_name_for_class_name(schema, class_name, macro_edge_name
 
 def _get_minimal_query_ast_from_macro_ast(macro_ast):
     """Get a query that should successfully compile to IR if the macro is valid."""
-    macro_directives = get_directives_for_ast(macro_ast)
     query_ast = remove_directives_from_ast(macro_ast, {
         directive.name
         for directive in MACRO_EDGE_DIRECTIVES
     })
 
-    # Add an output directive to make the ast a valid query
-    if 'output' not in macro_directives:
-        output_directive = Directive(Name('output'), arguments=[
-            Argument(Name('out_name'), StringValue('dummy_output_name'))
-        ])
-
-        # Add an output to a uuid field at the beginning of the query
-        # HACK(bojanserafimov): It's possible that the type at the top doesn't have a uuid field.
-        #                       It's also possible that none of the types in the macro have a field
-        #                       that we can output. The proper solution is to bypass the rule that
-        #                       a query has an output.
-        field = None
-        top_level_selections = query_ast.selection_set.selections[0].selection_set.selections
-        for selection in top_level_selections:
-            if isinstance(selection, Field):
-                if selection.name.value == 'uuid':
-                    field = selection
-        if field is None:
-            top_level_selections.insert(0, Field(Name('uuid'), directives=[]))
-            field = top_level_selections[0]
-        field.directives.append(output_directive)
+    # Add this output directive to make the ast a valid query
+    output_directive = Directive(Name('output'), arguments=[
+        Argument(Name('out_name'), StringValue('dummy_output_name'))
+    ])
+
+    # Find or create a __typename field
+    field = None
+    top_level_selections = query_ast.selection_set.selections[0].selection_set.selections
```

Comment:
Let's not mutate the `query_ast` or the field objects within. While `remove_directives_from_ast` gives you a new AST object, it relies on structural sharing to minimize the amount of copying it does, so if you mutate underlying object, you can still cause serious issues for other ASTs that rely on the same information.

Reply:
I am fairly certain that the above copying is not enough to prevent mutating the `query_ast` object. The copies give you a new `query_ast` object, and a new `SelectionSet` object within, which holds a reference to the same `selections` list within, which has the same objects inside etc. So, this `top_level_selections` list that we're mutating here is within one of the entries of that `selections` list, i.e. it's shared with `query_ast`.